### PR TITLE
Fix some leaks in systhreads

### PR DESCRIPTION
**Pull request by gasche**
This PR, joint work with @Engil and @fabbing, fixes some leaks in threads that we hope are the leak issues reported in #11289.

@toots, could you try on trunk (or 5.0) and tell us if you still observe leaks with this PR?

We wonder who would be interested in reviewing this. Maybe @gadmm, @xavierleroy or @sadiqj?

### CODE REVIEW

**Comment by gadmm** [comment]
I had a quick look, overall this looks good and makes sense to me. Not a formal review yet but let me know if you need one from me. I am also reassured that there is no overlap with the other PRs on systhreads.

The issue was originally for 4.13, is the fix substantially different there? Do we know what introduced the issue?

**Comment by gasche** [comment]
> Not a formal review yet but let me know if you need one from me.

If you are available and interested to do a review, then yes a review would be welcome. (I think it could definitely help other people decide to approve the PR.)

We haven't investigated the 4.13 systhreads codebase, so I don't know where leaks are there and how 4.14 fixed them. My guess is that the two implementations are very different, and the sources of leak are probably different as well.

**Comment by toots** [comment]
Thanks for working on this y'all. I'm happy to report that the leak seems fixed on my end as well:

Memory

File: runtime/backtrace_byt.c
Comment (Octachron) [inline runtime/backtrace_byt.c:269]: The allocation code above is still using `sizeof(code_t)`, should it be switched to `bactrace_slot` too?

File: runtime/domain.c
Comment (Octachron) [inline runtime/domain.c:1745]: This part is cleaning up some resource on the main domain thread and is not strictly speaking part of the systhread leak, isn't it?

Comment (gasche) [reply runtime/domain.c:1745]: Yes indeed. (This is a separated fourth commit.) We thought it made more sense to employ the new resource-cleanup in both places where they initialized, for threads *and* for domains. (Otherwise the code ends up in an arguably inconsistent state.)

File: runtime/backtrace_byt.c
Comment (gasche) [reply runtime/backtrace_byt.c:269]: Currently, `code_t` is defined as `opcode_t *`, and `backtrace_slot` as `void *` (both in caml/misc.h). Using `sizeof` on either ones gives the same result as they are both pointer types.

I think the code may be nicer if `backtrace_slot` was defined conditionally depending on whether we are in native or bytecode mode, so that it is *synonymous* to `code_t` in this file. In this alternative world, the change you suggest would not be necessary.

In the context of the present PR, we went for the simplest modification that worked without forcing us to write two different prototypes for the function. So we left `code_t` everywhere else in this file. Does that make sense?

File: runtime/fiber.c
Comment (gasche) [inline runtime/fiber.c:597]: Note: @Octachron and myself discussed this offline and we decided that it would be nice to have a documentation comment that explains what `gc_regs` and `gc_regs_buckets` are. I may submit this as a separate (complementary) PR.

File: runtime/domain.c
Comment (xavierleroy) [reply runtime/domain.c:1745]: +1 for cleaning up memory at domain exit too, not just thread exit.  The title of this PR and the Changes entry could reflect this fact.

File: runtime/fiber.c
Comment (xavierleroy) [reply runtime/fiber.c:597]: Yes, some documentation about `gc_regs_buckets` would be great.  I know what `gc_regs` does in OCaml 4, but I'm a bit lost as to the OCaml 5 implementation of `gc_regs`, what with the allocation via malloc and so on.

**Review by xavierleroy — APPROVED** [review]
Looks good to me overall, but I don't claim I understand the `gc_regs_bucket` business.  Let's wait for @Octachron's approval before merging.

File: runtime/backtrace_byt.c
Comment (Octachron) [reply runtime/backtrace_byt.c:269]: Keeping the change as small as possible makes sense.

File: runtime/fiber.c
Comment (gasche) [reply runtime/fiber.c:597]: For the record, I proposed some documentation of `gc_regs` and `gc_regs_buckets` in #11437.

**Comment by Octachron** [comment]
As far as I understand, `caml_thread_reinitialize` should also free the same resource as `caml_thread_remove_info` when it frees the list of threads within the domain.

**Comment by gasche** [comment]
> As far as I understand, `caml_thread_reinitialize` should also free the same resource as `caml_thread_remove_info` when it frees the list of threads within the domain.

My understanding is rather that only "pthread resources" (mutexes and what not) need to be reset after a fork, not memory which we assume was correctly duplicated. (A hint is that there is currently no code to cleanup the call stack in `reinitialize`, while there is one in `remove_info`.)

**Comment by Octachron** [comment]
The `reinitialize` function is currently freeing the current stack too? Moreover, the function is also freeing the thread information structure, thus there is no pointer left pointing to the `gc_reg_buffers` or `backtrace_buffer` structures for secondary threads. What I am missing?

**Comment by gasche** [comment]
Ah, right, the `reinitialize` function appears to be trying to cleanup all threads that are not the active thread. I don't know what the systhreads+fork combination is supposed to be doing, but clearly you are right and it should share the cleanup logic. Sorry for missing this earlier.

**Comment by gasche** [comment]
@Octachron this should have been addressed in cacc1cd61110f34bfe9d63a950c58e509c45dc98. Thanks!
( @Engil I wonder if you could confirm that this is the right approach. )

**Review by Octachron — APPROVED** [review]
I agree that this PR fixes a leak.

**Comment by gasche** [comment]
Merged, and cherry-picked in 5.0 as 29e912c24ad4a91eb00dde1bb78a9640c4ab0bb3. I had to resolve a conflict coming from #11390 (which is not (yet?) in 5.0).

**Comment by abbysmal** [comment]
Sorry, I'm currently on leave (will be back on monday).
I think the change is correct, however I spotted one thing I think is strange?

**Pull request merged**
